langchain/output_parsers/base.py

```diff
@@ -26,3 +28,22 @@ def dict(self, **kwargs: Any) -> Dict:
         output_parser_dict = super().dict()
         output_parser_dict["_type"] = self._type
         return output_parser_dict
+
+
+class OutputGuardrail(Guardrail, BaseModel):
```

Comment:
ah this was meant to be named OutputParserGuardRail

but yes, i can easily imagine some input guardrails

Reply:
How are you thinking about the scope of responsibility of an OutputGuardrail?

1) Take string output
2) Decode it (e.g., json decode)
3) Validate it
4) Any additional control logic? 
   -- e.g., logic to accept/reject the output if it looks like it contains bad language?
   -- retry logic on failure (i assume that should be at a higher level?)

----

Do you have a sense of how/why an InputGuardRail be different? (with the exception of potentially steps (1) and (2) being optional?)

Reply:
yeah i guess they're probably pretty interchangeable...

Reply:
the naming here is more reflective of the fact that it uses OutputParsers

Reply:
Agree with what I think @eyurtsev is getting at above:

The scope of OutputGuardrail/OutputParserGuardrail/whatever not super clear and seems not required right now: supposing Guardrail means some encapsulation of a validation/verification step + optional retry step. actually the Parser already raises validation error. and then the retry here will be very generic (likely, need to re-run the full forward pass again.) here's an example how this would look: https://github.com/hwchase17/langchain/blob/6628230a8a527b3f75abc529ecf80c5d6071b855/parsing_guardrails.ipynb

Reply:
How are various forms of validation + retry incorporated. It seems we these 2 cases. Pseudocode:

```
def predict(prompt):
  completion = llm(prompt)
  try:
    parsed_completion = parser(completion)
  except ParserException as e:
    # Handle failures captured by ECF
    if retry_parse:
      retry_completion = generic_retry_thing(prompt, completion)
      parsed_completion = parser(retry_completion)

  completion = parsed_completion
  
  for guardrail in self.guardrails:
    # Handle failures captured by guardrail
    completion = guardrail.evaluate(prompt, completion)

  return completion
```